### ISSUE TITLE
Various bugs in NTSC.cpp leading to incorrect hues

### ISSUE THREAD

**Issue by benrg** [AppleWin/AppleWin#1166]
It's been noted that AppleWin's lores color 1 is very purplish compared to other emulators. As far as I can tell this is due to bugs/design flaws in the NTSC code.

The code cites [this page](http://www.kreativekorp.com/miscpages/a2info/munafo.shtml) (by Robert Munafo) as a reference. According to that page, lores color 1 should have a UV phase of 90 (IQ phase of 33) and amplitude of 60/100. From that I get RGB coordinates (.93, -.10, .25). After clipping and applying the NTSC gamma of 2.2, I get (.86, 0, .05). That's NTSC RGB, not sRGB, but [comparing the gamuts](https://www.displaymate.com/Display_Color_Gamuts_2_files/image006.jpg) it's clear the color should be close to sRGB's \#f00 in hue.

There are several problems with the code. First, the 33° offset isn't added to the IQ phase anywhere; the phase is started at 45° (mod 90°). Second, the samples are doubled (from 560 per line to 1120) and the two half-samples are added at 45° increments, which effectively adds another 22.5°, except it's not exactly that because the amplitudes of the values that are added aren't equal. Third, the chroma amplitude is passed through two IIR filters, adding a lot of delay.

Fourth, the IIR filters aren't reinitialized between iterations of the loop. The signal and luma filters are damped enough that I think it doesn't matter, but the chroma filter has a long memory, so the previous bit pattern has an effect on the current bit pattern's chroma.

There's a comment in the source suggesting that the 45° start point (`CYCLESTART`) was chosen as a fudge factor to compensate for the other phase shifts, but I don't think any fudge factor works here. There's something screwy about the whole design. Even if you feed it a pure sine wave at the chroma frequency, the phase and amplitude never stabilize.

I don't know how real TVs of the period did it, but perhaps what you need is to replace the `initFilterChroma` oscillator with a 2D IQ-plane oscillator, and apply the phase to the input of it, instead of the output.

---

I think there are some mistakes in Munafo's page. First, I don't understand why the chroma amplitudes are 60 and 100. The ratio between these should be √2, presumably, so they should perhaps be 71 and 100. This doesn't significantly affect the hue of lores color 1. Second, he says conversion to RGB "requires a gamma correction for display on an RGB monitor", but the NTSC standard says that the RGB values obtained by a linear transformation from YIQ are already gamma corrected (with a nominal gamma of 2.2).

Another issue, as I mentioned above, is that NTSC RGB is significantly different from sRGB. Strictly speaking you should apply the NTSC gamma factor, convert to linear sRGB, then apply the inverse of the sRGB transfer function. The trouble is what to do with out-of-gamut values. Working out the best sRGB approximation to a color not in sRGB is hard.

**Comment by tomcw** [comment]
Hi @benrg,

>It's been noted that AppleWin's lores color 1 is very purplish compared to other emulators.

Where has this been noted? (I think it's the first I've heard about this.)

>The code cites [this page](http://www.kreativekorp.com/miscpages/a2info/munafo.shtml) (by Robert Munafo) as a reference.

I checked the original code for AppleWin's NTSC implementation (submitted as a source code zip by Sheldon Simms in 2011), and it doesn't contain this reference. So someone from the AppleWin team must have added this reference after refactoring the code for inclusion in AppleWin. (GitHub's "blame" says it was me... but I'm very sure it wasn't!)

So I don't think there's any link between the AppleWin NTSC implementation at this Munafo reference.

>There's a comment in the source suggesting that the 45° start point (CYCLESTART) was chosen as a fudge factor

I think you are referring to this:
https://github.com/AppleWin/AppleWin/blob/597cea1d86bc5dfe697a136c2aca75a497d915ae/source/NTSC.cpp#L930

It's possible that when one of us (AppleWin team members) refactored Sheldon's original code, that we changed the behaviour of something.
The original code (wsvideo.cpp) can be viewed [here](https://github.com/AppleWin/AppleWin/commit/22abd2b7a133dfed3c5cd8d1085fb042395796ba#diff-d06418ba69bd0b819dff867d5a61b461600208631b58c8ffc99d2c9317dac099).
(Sorry it's a diff - I don't know how to reference an old revision of a deleted file.)

If you want to suggest some code changes to improve things, then feel free to submit a PR (pull request).
NB. Please see the [CONTRIBUTING](https://github.com/AppleWin/AppleWin/blob/master/CONTRIBUTING.md) doc.

**Comment by Michaelangel007** [comment]
> So someone from the AppleWin team must have added this reference after refactoring the code for inclusion in AppleWin.

Yes, I did when I noticed Sheldon's colors were _slightly_ off.  I documented this in #254, specifically see this [comment](https://github.com/AppleWin/AppleWin/issues/254#issuecomment-67571076) years ago.

If you examine the first [commit](https://github.com/AppleWin/AppleWin/commit/e681c97c3b2fe94b94673dfe1e061a75050d3bee#diff-95840cf3e2f96432ed4cda977550603f784446063ca3f30dd74466f608c3189bR424) when `NTSC.cpp` was introduced you'll see I added a couple of comments when I noticed Sheldon's colors weren't _quite_ correct.

```c
#if CHROMA_BLUR
					//z = z * 1.25;
					zz = signal_prefilter(z);
					c = chroma_filter(zz); // "Mostly" correct _if_ CYCLESTART = PI/4 = 45 degrees
					y0 = luma0_filter(zz);
					y1 = luma1_filter(zz - c);
#else // CHROMA_BLUR
					y0 = y0 + (z - y0) / 4.0;
					y1 = y0; // fix TV mode

	#if CHROMA_FILTER == 0
					c = z; // sharper; "Mostly" correct _if_ CYCLESTART = 115 degrees
	#endif // CHROMA_FILTER
```
I never could get his code to match my monitor.

It didn't help that I didn't understand his phase initialization code. :-/

>  is that NTSC RGB is significantly different from sRGB. Strictly speaking you should apply the NTSC gamma factor, convert to linear sRGB, then apply the inverse of the sRGB transfer function

Yes, I wouldn't doubt that there lots of gamma conversion issues reside.

Sadly, I'm not a signal engineer.  Thankfully @sicklittlemonkey was able to provide some context years ago:

> About Marc Ressl's
> From http://www.digitalfaq.com/forum/video-restore/2944-ntsc-pal-video.html
> "Regarding the filter: It's a simple Lanczos lowpass filter with variable cutoff.
